datalad/support/protocol.py

```diff
@@ -0,0 +1,319 @@
+        pass
+
+    @property
+    def records_ext_commands(self):
+        return False
+
+    @property
+    def records_callables(self):
+        return False
+
+    @property
+    def do_execute_ext_commands(self):
+        return True
+
+    @property
+    def do_execute_callables(self):
+        return True
+
+
+class DryRunProtocol(ProtocolInterface):
+    """Protocol for dry runs.
+
+    Neither callables nor external commands are executed, when using this
+    protocol. Each recorded call results in a section, which is a dictionary
+    containing only the key 'command'. Its value is the list passed to
+    start_section() or add_section() respectively.
+    """
+    def __init__(self):
+        super(DryRunProtocol, self).__init__()
+        self._title = "Dry run protocol:" + linesep
+
+    def start_section(self, cmd):
+        id_ = len(self._sections)
+        self._sections.append({'command': cmd})
+        return id_
+
+    def end_section(self, id_, exception):
+        pass
+
+    def add_section(self, cmd, exception):
+        self.start_section(cmd)
+
+    @property
+    def records_callables(self):
+        return True
+
+    @property
+    def records_ext_commands(self):
+        return True
+
+    @property
+    def do_execute_callables(self):
+        return False
+
+    @property
+    def do_execute_ext_commands(self):
+        return False
+
+
+class DryRunExternalsProtocol(DryRunProtocol):
+    """Protocol for dry runs of external commands only.
+
+    Same as DryRunProtocol, but only affects (and records) external command
+    calls.
+    """
+    def __init__(self):
+        super(DryRunExternalsProtocol, self).__init__()
+
+    @property
+    def records_callables(self):
+        return False
+
+    @property
+    def do_execute_callables(self):
+        return True
+
+
+class ExecutionTimeProtocol(ProtocolInterface):
+    """Protocol to record execution times of callables as well as of external
+    commands.
+
+    Each recorded call results in a section, which is a dictionary
+    containing the keys 'command', 'start', 'end', 'duration' and 'exception'.
+    The value of 'command' is the list passed to start_section() or
+    add_section() respectively. The value of 'exception' is the instance of the
+    exception raised by the call or `None`. Times are stored as floating point
+    value in seconds.
+    """
+
+    def __init__(self):
+        super(ExecutionTimeProtocol, self).__init__()
+        self._title = "Execution time protocol:" + linesep
+
+    def start_section(self, cmd):
+        t_start = time.time()
+        id_ = len(self._sections)
+        self._sections.append({'command': cmd, 'start': t_start})
+        return id_
+
+    def end_section(self, id_, exception):
+        t_end = time.time()
+        self._sections[id_]['end'] = t_end
+        self._sections[id_]['duration'] = t_end - self._sections[id_]['start']
+        self._sections[id_]['exception'] = exception
```

Comment:
see -- and that is a typical use case for a context handler where upon entry it gets a new _id, yields, then upon continuation assigns... no need to expose that id outside